### DISCUSSION TITLE
Dealing with cascade layers

### DISCUSSION THREAD

**Discussion by o-t-w** [tailwindlabs/tailwindcss#6694] [Ideas]
Cascade layers are available since Chrome 99 and Firefox 97. As of March 8th it will be in Safari stable.

Cascade layers will be useful for Tailwind users because it will make it less necessary for users to set `important: true,` in their config as layers provide a nicer way to make sure utility classes always win.

However, for people that want the override styles provided by Tailwind, there will be a problem. 

A lot of people won't rely 100% on the Tailwind way of doing things, and will instead also want to write some of their own CSS. If I want to write any of my only styles inside a layer, these will be overridden by the Tailwind reset.
e.g.
```
@layer base {
.btn {
  background-color: rgb(200,200,200);
}
}
```
will be overridden by `button {background-color: transparent; }` from the Tailwind stylesheet. (Styles outside a layer will always win out over styles defined in a layer, regardless of specificity or source order.)

I think layers are a useful addition to CSS that will quickly gain widespread adoption. It would be awesome if Tailwind had a nice way of dealing with layers as soon as they gain cross-browser support.

**Comment by adamwathan** [comment]
Hey! The example you gave will already work, Tailwind reserves the `base`, `components`, and `utilities` layers and will merge those in to the right place automatically:

https://tailwindcss.com/docs/adding-custom-styles#using-css-and-layer

Your `.btn` class will not be overridden by `button {background-color: transparent; }`, here's a demo:

https://play.tailwindcss.com/vsxrhjhNcs?file=css

**Reply by o-t-w** [reply]
By polyfilling the feature you end up with: "Don't use `@layer` in component styles... You lose the ability to control the precedence of your styles, but unfortunately that’s totally out of our control because of how these tools work."

**Reply by o-t-w** [reply]
I don't think your response really addressed the issue. Your example is all Tailwind code. What if I had defined, _outside of my Tailwind CSS_, this CSS:

```
@layer overrides {
     .list {
  list-style-type: square;
}
}
```
This would be overridden by this Preflight code:
```
ol,
ul {
  list-style: none;
  margin: 0;
  padding: 0;
}
```
I know I could simply opt not to use Preflight and still use the utility classes without issue, but ideally Tailwind should have a config option to support native @layer. 
Because CSS that is not defined in a layer overrides all CSS defined in a layer, it basically makes it impossible to use native CSS `@layer` in conjunction with Preflight.
Tailwind needs a way to interact with non-Tailwind CSS code that is making use of the native browser `@layer` (rather than the Tailwind pollyfill version of `@layer`).
And I think the fact that Tailwind has its own version of `@layer` that is different from the browser CSS technology is just going to confuse everyone.

**Reply by o-t-w** [reply]
Some useful links:
- https://css-tricks.com/css-cascade-layers/#aa-using-design-systems-and-component-libraries
- https://twitter.com/bramus/status/1500223729309257737?s=21

**Comment by o-t-w** [comment]
Now that CSS `@layer` is supported in all browsers, it would be great to be able to define the Preflight into a separate layer from the utility classes.  Some sort of option in the config file to specify named layers for the utility classes and preflight would be much appreciated

**Comment by adamwathan** [comment]
Can you give me a real example of code that won’t work properly? I’m struggling to think of a situation that presents a real problem because you have complete control over where Preflight is inserted, and you can add any custom layers you want either before or after it. The whole reason Tailwind generates its code into three distinct layers is to give you this control.

**Reply by o-t-w** [reply]
The main use case is people that actually write their own CSS. Personally I love tailwind but rarely use it for 100% of styling. In cases where I am writing my own styling I do want to use @layer to help handle specificity. 
"you have complete control over where Preflight is inserted" - `@layer` ignores source order, so including it first doesn't solve the issue.

**Reply by o-t-w** [reply]
Here’s some styling from Preflight:
```
    *,
    ::before,
    ::after {
      border-width: 0;
      border-style: solid;
      border-color: theme('borderColor.DEFAULT', currentColor);
    }
    
    button,
    input,
    optgroup,
    select,
    textarea {
      font-family: inherit; /* 1 */
      font-size: 100%; /* 1 */
      font-weight: inherit; /* 1 */
      line-height: inherit; /* 1 */
      color: inherit; /* 1 */
      margin: 0; /* 2 */
      padding: 0; /* 3 */
    }
    
    button,
    [type='button'],
    [type='reset'],
    [type='submit'] {
      -webkit-appearance: button; /* 1 */
      background-color: transparent; /* 2 */
      background-image: none; /* 2 */
    }
```

Maybe I have my own hand-written styling somewhere, using @layer:

```
    @layer components {
    .button {
      color: white;
      border: solid 2px red;
      background-color: blue;
    }
    }
  ```

None of the `.button` styles are applied, regardless of source order. The `.button` class is setting the border-width to two pixels. That will be overridden by the `border-width: 0;` style defined in Preflight, regardless of source order. `color: white` is overridden by `color: inherit` and `background-color: blue;` is overridden by `background-color: transparent`.

**Reply by wongjn** [reply]
One could manually wrap the `@tailwind` directives in `@layer`s:

```css
@layer tailwind.base {
  @tailwind base;
}

@layer tailwind.components {
  @tailwind components;
}

@layer tailwind.utilities {
  @tailwind utilities;
}
```

See [example play](https://play.tailwindcss.com/Cev2wY3f6p?file=css). You could then declare your layer order and have more layers elsewhere etc.

**Reply by o-t-w** [reply]
This would be relatively easy for you to implement. All Tailwind needs to do is wrap the preflight code in a @layer:

```
@layer preflight {
    /* Preflight code here */
}
```

**Reply by o-t-w** [reply]
@wongjn That a great solution, thanks. Where is that code meant to go? I am using Tailwind in a Remix project. My `.css` file is built by the build tool - I don't hand code it. I include my own styling in a separate file. I'm looking in the Tailwind docs and it's not clear how where this code is meant to go.

**Reply by wongjn** [reply]
As per the [Remix documentation on Tailwind integration](https://remix.run/docs/en/v1/guides/styling#tailwind-css):

> If you want to use Tailwind's `@apply` method to extract custom classes, create a css file in the root directory, eg `./styles/tailwind.css`:
>
> ```css
> @tailwind base;
> @tailwind components;
> @tailwind utilities;
>
> @layer components {
>   .custom-class {
>     @apply ...;
>   }
> }
> ```
> Then alter how Tailwind is generating your css:
>
> ```diff
>  {
>   // ...
>   "scripts": {
>     "build": "run-s \"build:*\"",
>     "build:css": "npm run generate:css -- --minify",
>     "build:remix": "remix build",
>     "dev": "run-p \"dev:*\"",
>     "dev:css": "npm run generate:css -- --watch",
>     "dev:remix": "remix dev",
> +   "generate:css": "npx tailwindcss -i ./styles/tailwind.css -o ./app/tailwind.css",
>     "start": "remix-serve build"
>   }
>   // ...
>  }
> ```

So piggy-backing off this, you would edit the above `./styles/tailwind.css` to be like:

```css
@layer tailwind.base {
  @tailwind base;
}

@layer tailwind.components {
  @tailwind components;
}

@layer tailwind.utilities {
  @tailwind utilities;
}
```

**Reply by adamwathan** [reply]
@o-t-w Sorry can you give me a _full_ example? Like a full CSS file that includes your custom styles and Tailwind's styles so I can actually understand. When you give me this code snippet:

```css
    @layer components {
    .button {
      color: white;
      border: solid 2px red;
      background-color: blue;
    }
    }
```

...I don't know where you are actually _putting_ that code.

Tailwind does reserve the `base`, `components`, and `utilities` layers but if you come up with your own layer names you can really do whatever you want, you just need to make sure your layers are defined in the location they should be.

For example:

```css
@tailwind base;
@layer app-base;

@tailwind components;
@layer app-components;

@tailwind utilities;
@layer app-utilities;
```

Now you can write this and it definitely _will_ come after your base styles so it will override Preflight just fine:

```css
@layer components {
  .button {
     color: white;
     border: solid 2px red;
     background-color: blue;
  }
}
```

Can you give an example where this strategy doesn't work? That's what I'm trying to understand. I feel like we are doing a lot already to make Tailwind very compatible with custom CSS that use a layered approach — that's the whole reason we split things into the three `@tailwind` declarations, so you can insert stuff in between.

**Reply by o-t-w** [reply]
Thank you @wongjn this solution works amazingly well! :) 
@adamwathan I guess I can't "close" this idea the way I can close a GitHub issue, but my use case is already solved by @wongjn without needing to add anything to Tailwind. I do think it might be worth adding something to the documentation about this though as I'm sure other people will have the same question as cascade layers become more widely used now that browser support is pretty good. Although from your last response I still feel like we are talking past each other. I'm happy to do a PR to the docs. 

And I do think it is massively unfortunate that Tailwind is using the same syntax as a native CSS feature. It is bound to cause loads of confusion.

**Reply by o-t-w** [reply]
@adamwathan I have pushed a simple example to GitHub. https://github.com/o-t-w/cascade-layers-tailwind
Tailwind CSS source file is in top level /styles directory. My own custom CSS is written in styles.css file in the app directory.

**Reply by adamwathan** [reply]
Was just looking into this a bit more and I understand it better now — the thing I didn't know before is that any CSS that's _not_ in a `@layer` is automatically moved _after_ any declared layers, so any custom user layers would come before Tailwind's rules even if you tried to declare the layer order yourself in a specific way. Didn't know that until today, sorry!

So yeah @wongjn's solution is definitely the way to go, glad that works and that there's nothing in Tailwind expecting those rules to all be at the root of the document 👍🏻 Will be sure to make sure that continues to work going forward.

**Reply by nemzes** [reply]
I don't think the example from @wongjn fixes this issue cleanly. The styles provided by `@tailwind utilities` leak outside the CSS `@layer` in the generated CSS in their example.

**Reply by o-t-w** [reply]
I documented the approach I went for on CSS Tricks: 
https://css-tricks.com/using-css-cascade-layers-to-manage-custom-styles-in-a-tailwind-project/

**Reply by nemzes** [reply]
I just saw your article @o-t-w; great write-up! I think there is a problem still, if you see the "generated CSS" in [@wongjn's play](https://play.tailwindcss.com/Cev2wY3f6p?file=css) there are some rules that leak at the end of the `@layer` wrappers. They seem to be pseudo-classes and media queries, which should remain within their `@layer`s.

![image](https://user-images.githubusercontent.com/145197/186680426-a038016d-c286-4dfd-9587-cb380f252151.png)

**Reply by o-t-w** [reply]
@nemzes those are styles that come from `@tailwind variants`. If you explicitly include `@tailwind variants;` hopefully it should be fine?

**Reply by nemzes** [reply]
Indeed. Adding an extra `@layer` seems to fix it, although I was surprised that those styles just popped up "magically" (I see the logic in inserting variants at the end, but it makes this solution feel somewhat brittle given the assumptions that Tailwind makes don't necessarily aling with `@layer`):

```css
@layer tailwind.variants {
  @tailwind variants;
}
```

**Comment by nemzes** [comment]
Even if it's possible to deal with the override problem (see [@wongjn's solution above](https://github.com/tailwindlabs/tailwindcss/discussions/6694#discussioncomment-3340814)), it's not clear to me how I can use Tailwind's `@layer` feature together with CSS Cascade Layers' `@layer` — since they name-clash.

Won't the Tailwind processor simply assume that `@layer` declarations refer to Tailwind's own layer feature? Should Tailwind's `@layer` be renamed `@tailwind-layer` instead?

**Reply by o-t-w** [reply]
I’ve not noticed any issues, but agree it feels slightly hazardous…

**Reply by adamwathan** [reply]
Tailwind reserves the “base”, “components”, and “utilities” layer names but everything else is untouched and just passed through to your final CSS. So if you are hellbent on using those layer names in your custom CSS you’re out of luck but feels like a very superficial problem to me that’s not worth getting too annoyed about. We actually had `@layer` before it existed in CSS, otherwise probably would have used a different name but don’t want to mess with backwards compatibility.

**Comment by nemzes** [comment]
It seems like the solution mentioned above will trigger the `Nested @tailwind rules were detected, but are not supported` warning, courtesy of a [parent node check](https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/detectNesting.js#L8).

**Comment by trey** [comment]
I'm still getting that nesting error, but things seem to be working as expected. Thanks for that article, @o-t-w!

**Comment by GuusPlx** [comment]
I think it's best if Tailwind doesn't polyfill the `@layer` feature anymore, or provides a differnt approach to have better compatability with the default `@layer` CSS behaviour.

@adamwathan here is a real example of code that won't work:

Currently it is not possible to dynamically import multiple css files (for code splitting for example) and use the layer functionality to create the right CSS layering.
Tailwind layer polyfill works great for one CSS bundle. But when you want multiple bundles that are dynamically imported on different routes it won't work. 

``` html
<head>
  <!--  (includes base styles and tailwind) -->
  <link rel="stylesheet" href="main.css"> 
  
  <!--  these files can't use cascading layer functionality to get onto the right layer -->
  <link rel="stylesheet" href="dashboard-bundle.css"> <!--  (optionally rendered based on route) -->
  <link rel="stylesheet" href="other-large-module.css"> <!--  (optionally rendered based on route) -->
</head>
```

I like the polyfill TailwindCSS currently uses through PostCSS but it should also work together with the default `@layer` CSS functionality that is currently supported in all browsers

https://css-tricks.com/css-cascade-layers/#aa-using-design-systems-and-component-libraries
The approach of this article doesn't seem to work for me and causes this error:
`"Nested @tailwind rules were detected, but are not supported."`

What do you think of this situation? It is mostly caused because in our projects we use CSS together with utility classes of TailwindCSS so we do depend on large CSS modules for specific pages. 
Is it a valid case or can it be solved any other way?
Thanks in advance!

**Reply by adamwathan** [reply]
Yep planning to change this for v4, but can’t make any breaking changes until then. The next version of Tailwind will support wrapping our “layers” in real cascade layers like you’ve mentioned though — try our insiders build if you’d like to test.

**Reply by GuusPlx** [reply]
Awesome Adam! Tailwind just keeps getting better, thanks alot.

**Reply by DaniGuardiola** [reply]
@adamwathan could you link to any docs or examples of how to do that in the insiders build? Would love to beta-test :)

**Reply by TKDev7** [reply]
actually where can I find this insiders build?
